morpheus/io/utils.py

```diff
@@ -31,3 +39,67 @@ def filter_null_data(x: DataFrameType):
         return x
 
     return x[~x['data'].isna()]
+
+
+def _cudf_needs_truncate(df: cudf.DataFrame, column_max_bytes: dict[str, int]) -> bool:
+    """
+    Optimization, cudf contains a byte_count() method that pandas lacks.
+    """
+    for (col, max_bytes) in column_max_bytes.items():
+        series: cudf.Series = df[col]
+
+        assert series.dtype == 'object'
+
+        if series.str.byte_count().max() > max_bytes:
+            return True
+
+    return False
+
+
+def truncate_string_cols_by_bytes(df: DataFrameType,
+                                  column_max_bytes: dict[str, int],
+                                  warn_on_truncate: bool = True) -> DataFrameType:
+    """
+    Truncates all string columns in a dataframe to a maximum number of bytes.
+
+    If truncation is not needed, the original dataframe is returned. If `df` is a cudf.DataFrame, and truncating is
+    needed this function will convert to a pandas DataFrame to perform the truncation.
+
+    Parameters
+    ----------
+    df : DataFrameType
+        The dataframe to truncate.
+    column_max_bytes: dict[str, int]
+        A mapping of string column names to the maximum number of bytes for each column.
+    warn_on_truncate: bool, default True
+        Whether to log a warning when truncating a column.
+
+    Returns
+    -------
+    DataFrameType
+        The truncated dataframe, if needed.
+    """
+
+    if isinstance(df, cudf.DataFrame):
+        # cudf specific optimization
+        if not _cudf_needs_truncate(df, column_max_bytes):
+            return df
+
+        # If truncating is needed we need to convert to pandas to use the str.encode() method
+        df = df.to_pandas()
```

Comment:
Instead of converting the entire dataframe, just convert the string columns as needed. And return the same dataframe type that was pass in the args